### PR TITLE
Fix UI so it runs properly with modern urwid

### PR DESCRIPTION
**Pull request by ogayot**
After upgrading to noble, I noticed several issues when running the Subiquity UI.

The version of `urwid` changed from 2.1.2-4 to 2.6.4-1. The change that impacts us most is that urwid dropped what was called the "idle emulation delay" [1]. Without it, urwid only draws the screen after it finishes handling an event (i.e., when the event handler returns).

This means that code like this:

```python
connect_signal(self.form, "submit", self.done)

def done(self):
    self._w = screen(...)
```
will still be followed by a screen redraw. However, for code like this:

```python
connect_signal(self.form, "submit", self.done)

def done(self):
    async def do_something():
        self._w = screen(...)
    run_bg_task(do_something())
```
the screen redraw only happens **before** the `self._w = screen(...)` statement is executed.
In the new urwid, there will be no additional call to `urwid.main_loop.draw_screen()` after the statement is executed. This means we have to call it ourselves.

To redraw the screen after an event happens outside of urwid, one can call the `urwid.main_loop.draw_screen()` function. This call should be done in the main thread only.

TIL urwid alarms
------------------

urwid supports timers called alarms. They can be used as follows:

```python
def callback(alarm, user_data):
     # TODO

urwid.main_loop.set_alarm_in(0.3, callback, user_data)
```
I decided not to use them but for use-cases where we want to update the UI after `n` seconds, they would have an edge over `asyncio.loop.call_later(callback)` and `await asyncio.sleep()` because they would redraw the UI automatically after the event occurs. Yes, they generate urwid events so the redraw rules applies.

implementation changes
-----------------------
* After a view is modified outside of a urwid even handler, one should call the `view.request_redraw_if_visisble()` to make the change visible. This will  schedule a screen redraw if the view is currently visible.
* I've added calls to `view.request_redraw_if_visible()` at all the places where I noticed the screen was not redrawing automatically.
* Calling spinner.start() does not call spinner.stop() anymore. Doing so was causing UI glitches because the spinner becomes invisible when .stop() is called but does not immediately become visible again when .start() is called.
* Spinners can take a `debug=True` argument to make them log events. This makes it easier to diagnose spinners that were never stopped and continue causing asyncio context switches and screen redraws

What's still broken on noble (AFAIK)?
-------------------
[x] When running integration tests, you will likely end up with the following crash:
```python
  File "/usr/lib/python3/dist-packages/urwid/widget/widget.py", line 401, in pack
    raise WidgetError(f"Cannot pack (maxcol,) size, this is not a flow widget: {self!r}")
urwid.widget.widget.WidgetError: Cannot pack (maxcol,) size, this is not a flow widget: <Padding widget <FocusTrackingPile widget> align=<Align.CENTER: 'center'> min_width=14 width=14>
```
This is addressed by https://github.com/urwid/urwid/pull/843 and is available in urwid 2.6.5. On noble we have 2.6.4-1 so it should be available soon.
[ ] `make lint` fails with:
``` python
File "/usr/lib/python3/dist-packages/flake8/plugins/finder.py", line 298, in _load_plugin
    raise FailedToLoadPlugin(plugin.package, e)
flake8.exceptions.FailedToLoadPlugin: Flake8 failed to load plugin "pycodestyle" due to cannot import name 'missing_whitespace_around_operator' from 'pycodestyle' (/usr/lib/python3/dist-packages/pycodestyle.py).
make: *** [Makefile:76: flake8] Error 1
```
This is addressed by https://launchpad.net/ubuntu/+source/python-flake8 currently stuck in proposed.

[1] https://github.com/urwid/urwid/pull/541

### CODE REVIEW

**Comment by ogayot** [comment]
Filed https://bugs.launchpad.net/ubuntu/+source/urwid/+bug/2058388 since it also affects the CI.

**Review by dbungert — COMMENTED** [review]
Thanks for sorting this out.

+1 for main branch when the known issues are resolved, let's specifically not merge this to `ubuntu/noble`.

**Comment by ogayot** [comment]
Rebased on top of main to drop the failing test on EOL lunar.

**Review by Chris-Peterson444 — APPROVED** [review]
This looks mostly good to me. Although, I am a little confused on difference between the scenarios in which a controller should call out to `view.request_redraw_if_visible` vs `app.request_screen_redraw`. I guess this should be case by case (i.e., if we know the screen will need a redraw we can request the redraw directly) ? Although I wonder how bad it would be if we just always redraw, instead of conditionally. Either way, this works for now. Thanks for putting the work into this! 

I additionally did some light testing of this on Noble with urwid 2.6.10-1 and it worked out well so far!

**Review by dbungert — APPROVED** [review]
I think this is ready when Chris' feedback is addressed, lints are passing now, and it could benefit from a rebase.

File: subiquity/client/controllers/ssh.py [hunk 1]
```diff
@@ -100,19 +100,22 @@ async def _fetch_ssh_keys(self, *, context, ssh_import_id):
                 self.ui.body.fetching_ssh_keys_failed(
                     _("Importing keys failed:"), response.error
                 )
+                self.ui.body.request_redraw_if_visible()
```
[the remark this answers is not in the document]
Comment (ogayot) [reply subiquity/client/controllers/ssh.py:103]: Right, that would work too and would avoid a double isinstance(). Thanks.

[the remark this answers is not in the document]
Comment (ogayot) [reply subiquity/client/controllers/ssh.py:103]: I still prefer to operate on the view itself so I've added a view.request_redraw() method. Hope that's okay with you.

**Pull request merged**